File: android/app/src/main/kotlin/com/iblocker/android/upstream/UdpUpstream.kt
package com.iblocker.android.upstream

import com.iblocker.core.engine.DnsUpstream
import com.iblocker.core.engine.UpstreamException
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.util.concurrent.ConcurrentHashMap
import kotlin.concurrent.thread

/**
 * Plain DNS-over-UDP:53 upstream.
 *
 * One long-lived connected socket serves every client, so transaction IDs are
 * rewritten to a private sequence and demultiplexed on receive. The address
 * must be an IP literal — hostname resolution inside the service would
 * recurse through the interface it is serving.
 */
class UdpUpstream(
    private val address: String,
    private val port: Int = 53,
    private val protector: SocketProtector = SocketProtector.NONE,
    private val timeoutMillis: Long = 3_000,
) : DnsUpstream {

    private val pending = ConcurrentHashMap<Int, CompletableDeferred<ByteArray>>()

    @Volatile
    private var socket: DatagramSocket? = null

    @Volatile
    private var closed = false

    private var nextID = 0
    private val idLock = Any()

    override suspend fun resolve(query: ByteArray): ByteArray {
        var lastError: Exception = UpstreamException.Timeout()
        repeat(2) {
            try {
                return attempt(query)
            } catch (error: Exception) {
                lastError = error
            }
        }
        throw lastError
    }

    private suspend fun attempt(query: ByteArray): ByteArray {
        if (query.size < 12) throw UpstreamException.BadResponse()
        val socket = ensureSocket()

        val bytes = query.copyOf()
        val originalID = ((bytes[0].toInt() and 0xFF) shl 8) or (bytes[1].toInt() and 0xFF)

        val ourID = synchronized(idLock) {
            var candidate = (nextID + 1) and 0xFFFF
            var attempts = 0
            while (pending.containsKey(candidate) && attempts < 0x10000) {
                candidate = (candidate + 1) and 0xFFFF
                attempts += 1
            }
            nextID = candidate
            candidate
        }
        bytes[0] = ((ourID shr 8) and 0xFF).toByte()
        bytes[1] = (ourID and 0xFF).toByte()

        val waiter = CompletableDeferred<ByteArray>()
        pending[ourID] = waiter
        try {
            withContext(Dispatchers.IO) {
                socket.send(DatagramPacket(bytes, bytes.size, InetAddress.getByName(address), port))
            }
            val response = withTimeout(timeoutMillis) { waiter.await() }
            if (response.size < 12) throw UpstreamException.BadResponse()
            val out = response.copyOf()
            out[0] = ((originalID shr 8) and 0xFF).toByte()
            out[1] = (originalID and 0xFF).toByte()
            return out
        } catch (_: TimeoutCancellationException) {
            throw UpstreamException.Timeout()
        } catch (error: Exception) {
            if (error is UpstreamException) throw error
            throw UpstreamException.ConnectionFailed(error.message ?: "send failed")
        } finally {
            pending.remove(ourID)
        }
    }

    @Synchronized
    private fun ensureSocket(): DatagramSocket {
        socket?.takeIf { !it.isClosed }?.let { return it }
        val created = DatagramSocket()
        protector.protect(created)
        socket = created
        startReceiveLoop(created)
        return created
    }

    private fun startReceiveLoop(socket: DatagramSocket) {
        thread(isDaemon = true, name = "iblocker-udp-upstream") {
            val buffer = ByteArray(4096)
            while (!closed && !socket.isClosed) {
                try {
                    val packet = DatagramPacket(buffer, buffer.size)
                    socket.receive(packet)
                    if (packet.length < 2) continue
                    val data = packet.data.copyOfRange(packet.offset, packet.offset + packet.length)
                    val id = ((data[0].toInt() and 0xFF) shl 8) or (data[1].toInt() and 0xFF)
                    pending.remove(id)?.complete(data)
                } catch (_: Exception) {
                    if (closed) break
                    // Socket died: drop it so the next query reconnects, and
                    // fail everyone waiting rather than leaving them to time out.
                    failAllWaiting("socket closed")
                    break
                }
            }
        }
    }

    private fun failAllWaiting(reason: String) {
        val waiting = pending.values.toList()
        pending.clear()
        waiting.forEach { it.completeExceptionally(UpstreamException.ConnectionFailed(reason)) }
        socket?.let { runCatching { it.close() } }
        socket = null
    }

    override fun close() {
        closed = true
        failAllWaiting("upstream closed")
    }
}

File: ios/IBlockerKit/Sources/IBlockerTunnelKit/UDPUpstream.swift
#if canImport(Network)
import Foundation
import Network
import IBlockerKit

/// Plain DNS-over-UDP:53 upstream.
///
/// One long-lived connected socket serves every client, so transaction IDs
/// are rewritten to a private sequence and demultiplexed on receive. The
/// address must be an IP literal — hostname resolution inside the tunnel
/// would recurse.
public actor UDPUpstream: DNSUpstream {
    private let host: NWEndpoint.Host
    private let port: NWEndpoint.Port
    private let timeoutNanoseconds: UInt64
    private var connection: NWConnection?
    private var nextID: UInt16 = 0
    private var pending: [UInt16: CheckedContinuation<Data, Error>] = [:]

    public init(address: String, port: UInt16 = 53, timeout: TimeInterval = 3) {
        self.host = NWEndpoint.Host(address)
        self.port = NWEndpoint.Port(rawValue: port) ?? 53
        self.timeoutNanoseconds = UInt64(timeout * 1_000_000_000)
    }

    public func resolve(_ query: Data) async throws -> Data {
        var lastError: Error = UpstreamError.timeout
        for _ in 0..<2 {
            do {
                return try await attempt(query)
            } catch {
                lastError = error
            }
        }
        throw lastError
    }

    private func attempt(_ query: Data) async throws -> Data {
        guard query.count >= 12 else { throw UpstreamError.badResponse }
        let connection = ensureConnection()

        var bytes = [UInt8](query)
        let originalID = UInt16(bytes[0]) << 8 | UInt16(bytes[1])

        nextID &+= 1
        var attempts = 0
        while pending[nextID] != nil && attempts < 0x10000 {
            nextID &+= 1
            attempts += 1
        }
        let ourID = nextID
        bytes[0] = UInt8(ourID >> 8)
        bytes[1] = UInt8(ourID & 0xFF)

        let response: Data = try await withCheckedThrowingContinuation { continuation in
            pending[ourID] = continuation
            connection.send(content: Data(bytes), completion: .contentProcessed { [weak self] error in
                if let error {
                    Task {
                        await self?.fail(id: ourID, error: UpstreamError.connectionFailed(String(describing: error)))
                    }
                }
            })
            Task { [weak self, timeoutNanoseconds] in
                try? await Task.sleep(nanoseconds: timeoutNanoseconds)
                await self?.fail(id: ourID, error: UpstreamError.timeout)
            }
        }

        var out = [UInt8](response)
        guard out.count >= 12 else { throw UpstreamError.badResponse }
        out[0] = UInt8(originalID >> 8)
        out[1] = UInt8(originalID & 0xFF)
        return Data(out)
    }

    private func ensureConnection() -> NWConnection {
        if let existing = connection {
            return existing
        }
        let conn = NWConnection(host: host, port: port, using: .udp)
        conn.stateUpdateHandler = { [weak self] state in
            switch state {
            case .failed(let error):
                Task { await self?.connectionBroke(String(describing: error)) }
            case .cancelled:
                Task { await self?.connectionBroke("cancelled") }
            default:
                break
            }
        }
        receiveLoop(conn)
        conn.start(queue: .global(qos: .userInitiated))
        connection = conn
        return conn
    }

    private nonisolated func receiveLoop(_ conn: NWConnection) {
        conn.receiveMessage { [weak self] data, _, _, error in
            guard let self else { return }
            if let data, !data.isEmpty {
                Task { await self.complete(data) }
            }
            if error == nil {
                self.receiveLoop(conn)
            }
        }
    }

    private func complete(_ data: Data) {
        guard data.count >= 2 else { return }
        let id = UInt16(data[0]) << 8 | UInt16(data[1])
        if let continuation = pending.removeValue(forKey: id) {
            continuation.resume(returning: data)
        }
    }

    private func fail(id: UInt16, error: Error) {
        if let continuation = pending.removeValue(forKey: id) {
            continuation.resume(throwing: error)
        }
    }

    private func connectionBroke(_ reason: String) {
        connection = nil
        let waiting = pending
        pending.removeAll()
        for continuation in waiting.values {
            continuation.resume(throwing: UpstreamError.connectionFailed(reason))
        }
    }
}
#endif
